docs/modules/ROOT/pages/how-tos/worklows/git-submodules.adoc

```diff
@@ -0,0 +1,128 @@
+
+The downstream repo can be used to do the following:
+
+* hold the git submodule reference to the upstream repository.
+* use the contents of the git submodule in a Containerfile for further edits.
+* onboarded onto {ProductName}.
+
+==== Configuring components
+
+Here is an example of a component definition:
+
+[source, yaml]
+----
+apiVersion: appstudio.redhat.com/v1alpha1
+kind: Component
+metadata:
+  name: <component-name>
+  namespace: <project-namespace>
+  annotations:
+    build.appstudio.openshift.io/request: configure-pac
+    image.redhat.com/generate: '{"visibility": "public"}'
+spec:
+  application: <application-name>
+  componentName: <component-name>
+  source:
+    git:
+      revision: main
+      url: <downstream repo name>
+      dockerfileUrl: <Dockerfile/Containefile path in downstream repository>
+      context: <directory path used to build the Containerfile/Dockerfile>
+----
+
+See xref:/how-tos/creating.adoc[Creating applications and components] for more details.
+
+==== Configuring the Containerfile
+
+In the Containerfile you can specify the git submodule directories and add any necessary edits that differ from the upstream repository. Here is a basic example:
+
+[source, Dockerfile]
+----
+FROM <base-image>
+ENV <SUBMODULEPKG> /path/to/submodule
+
+# Install necessary dependencies
+RUN dnf -y install <dependencies>
+
+# Set the working directory
+WORKDIR ${<SUBMODULEPKG>}
+
+RUN sed -i 's/<old-text>/<new-text>/g' /file/in/submodule/directory
+
+WORKDIR /app
+
+ENTRYPOINT ["<entrypoint-executable>"]
+----
+
+=== Create the git submodule for the upstream repo to track it
+
+After creating the downstream repo, add a git submodule for the upstream repo:
+
+[source, bash]
+----
+git submodule add <upstream-url>
+----
+
+Check and configure the git submodule config:
+[source, gitmodules]
+----
+# .gitmodules
+[submodule "<project-name>"]
+  path = <path-in-downstream-repo>
+  url = https://github.com/<namespace>/<project>.git
+----
+See the link:https://git-scm.com/docs/gitsubmodules[Git documentation] for more details on configuring git submodules.
+
+**Note**: By default the branch is not specified in the git submodule configuration file when adding a git submodule.
+see link:https://git-scm.com/docs/gitsubmodules[Git documentation] for more details on how to configure the branches for the submodules.
+
+=== Onboard the component onto {ProductName}
+
+After creating the downstream repository you can onboard the component. See xref:/how-tos/creating.adoc[Creating applications and components] for more details.
+
+=== Using private repositories as a git submodule
+
+==== Github
+
+When using private repositories we encourage that you give the link:https://github.com/apps/red-hat-konflux[Red Hat {ProductName} Github app] access to that private repository.
+
+==== Gitlab
+
+For Gitlab repositories please see xref:/how-tos/configuring/creating-secrets.adoc#creating-source-control-secrets[Creating secrets for your builds]
```

Comment:
The context for this is that users will need to configure the secret and then these private repositories should "just work"? Can we add that context to this section?

Reply:
I removed this section. This was meant more for general Gitlab support, not the "using private repos" scenario. I have not found a way to do this outside of the Github Konflux application. Maybe we can add a new section once we have a similar app on Gitlab?

Reply:
Ah, interesting. You haven't found a way for a downstream repo to be in GitLab or the submodule? 

(I guess we are talking about this same topic in two threads now).

Reply:
I haven't found a way to deal with private submodules on Gitlab, at this time. Until I can confirm that with a similar test like I ran for the Github repo, I think it's better to leave that out for now. On Gitlab I did get MRs to update my public submodules, but I still need to confirm the private ones.

I'll add it soon after confirming, my Gitlab repo is on Konflux and once it's fully configured and building I can run a test like I did for the private Github submodules.

Reply:
Should we add a caveat that at least states that GitLab submodules are untested (or at least private GitLab)?

Reply:
Yeah that's fine, let me add it now....

Reply:
Caveat added: https://github.com/konflux-ci/docs/pull/120/commits/8fd002b00b89825db0703c22ea161b6f258554df